app/app.go

```diff
@@ -240,6 +244,139 @@ func (a *app) ProvisionDevice(
 	return err
 }
 
+func (a *app) syncBatch(
+	ctx context.Context,
+	devices []model.Device,
+	integCache map[uuid.UUID]*model.Integration,
+) error {
+	// FIXME(alf): should be able to continue on (partial) error
+	var err error
+
+	deviceMap := make(map[uuid.UUID][]string, len(integCache))
+	for _, dev := range devices {
+		for _, id := range dev.IntegrationIDs {
+			deviceMap[id] = append(deviceMap[id], dev.ID)
+		}
+	}
+
+	for integID, deviceIDs := range deviceMap {
+		integration, ok := integCache[integID]
+		if !ok {
+			// (Data race) Try again to fetch the integration
+			integration, err = a.store.GetIntegrationById(ctx, integID)
+			if err != nil {
+				if err == store.ErrObjectNotFound {
+					integCache[integID] = nil
+				} else {
+					return err
+				}
+			} else {
+				integCache[integID] = integration
+			}
+		}
+		if integration == nil {
+			// Should not occur, but is not impossible since mongo client
+			// caches batches of results.
+			_, err := a.store.RemoveIntegrationFromDevices(ctx, integID)
+			if err != nil {
+				return errors.Wrap(err, "failed to remove integration from devices")
+			}
+			continue
+		}
+
+		switch integration.Provider {
+		case model.ProviderIoTHub:
+			err := a.syncIoTHubDevices(ctx, deviceIDs, *integration)
+			if err != nil {
+				return err
+			}
+		default:
+			// Invalid integration
+			// FIXME(alf) what to do?
+		}
+	}
+
+	return nil
+}
+
+func (a app) syncCacheIntegrations(ctx context.Context) (map[uuid.UUID]*model.Integration, error) {
+	// Cache integrations for the given tenant
+	integrations, err := a.store.GetIntegrations(
+		ctx, model.IntegrationFilter{Limit: 20},
```

Comment:
Why 20? By the way, maybe it should be a constant.

Reply:
Right now, we don't allow more than 1 anyway, so I just picked a "reasonable" number.